lib/edgehog_device_forwarder/websockets.ex

```diff
@@ -0,0 +1,166 @@
+  @type controller :: pid()
+  @type id :: Data.socket_id()
+  @type socket :: pid()
+  @type message ::
+          {:text, String.t()}
+          | {:binary, binary}
+          | {:ping, binary}
+          | {:pong, binary}
+          | {:close, ProtoWebSocket.Close.t()}
+
+  @spec upgrade(id, socket, Data.t()) ::
+          :ok | {:error, :http_request_not_found}
+  def upgrade(request_id, socket, socket_data) do
+    ConCache.isolated(@table, request_id, fn ->
+      case ConCache.get(@table, request_id) do
+        {:upgrading, :websocket, queue} ->
+          for message <- MessageQueue.to_list(queue) do
+            send(socket, message)
+          end
+
+          start_monitor(socket, socket_data)
+          ConCache.dirty_put(@table, request_id, {:websocket, socket})
+
+        nil ->
+          "upgrade: request_id #{inspect(request_id)} not found"
+          |> Logger.notice(tag: "invalid_http_id")
+
+          {:error, :http_request_not_found}
+
+        other ->
+          "upgrade: request_id #{inspect(request_id)}. expected {:upgrading, :websocket}, got #{inspect(other)}"
+          |> Logger.notice(tag: "invalid_http_id")
+
+          {:error, :http_request_not_found}
+      end
+    end)
+  end
+
+  @spec forward(id, message) ::
+          :ok | {:error, :websocket_is_closing} | {:error, :websocket_not_found}
+  def forward(socket_id, message) do
+    ConCache.isolated(@table, socket_id, fn ->
+      case ConCache.get(@table, socket_id) do
+        {:websocket, socket} ->
+          {type, _} = message
+
+          if type == :close do
+            ConCache.dirty_put(@table, socket_id, {:closing, :websocket})
+          end
+
+          send(socket, message)
+
+          :ok
+
+        {:upgrading, :websocket, queue} ->
+          queue = MessageQueue.add(queue, message)
+          ConCache.dirty_put(@table, socket_id, {:upgrading, :websocket, queue})
+
+          :ok
+
+        {:closing, :websocket} ->
+          {type, _} = message
+
+          if type == :close do
+            ConCache.dirty_delete(@table, socket_id)
+            :ok
+          else
+            """
+            fetch: socket_id #{inspect(socket_id)}:
+            websocket is closing, received: #{inspect(message, pretty: true)}
+            """
+            |> Logger.notice(tag: "invalid_websocket_id")
+
+            {:error, :websocket_is_closing}
+          end
+
+        nil ->
+          """
+          fetch: socket_id #{inspect(socket_id)} not found.
+          message received: #{inspect(message, pretty: true)}"
+          """
+          |> Logger.notice(tag: "invalid_websocket_id")
+
+          {:error, :websocket_not_found}
+
+        other ->
+          """
+          fetch: socket_id #{inspect(socket_id)}: expected :websocket, found #{inspect(other)}
+          message received: #{inspect(message, pretty: true)}
+          """
+          |> Logger.notice(tag: "invalid_websocket_id")
+
+          {:error, :websocket_not_found}
+      end
+    end)
```

Comment:
Consider moving this (or even just the happy path) to a private function to improve readability

Reply:
Done